src/ParallelAlgorithms/Amr/Criteria/Constraints.hpp

```diff
@@ -0,0 +1,220 @@
+#include "Options/Context.hpp"
+#include "Options/ParseError.hpp"
+#include "Options/String.hpp"
+#include "ParallelAlgorithms/Amr/Criteria/Criterion.hpp"
+#include "Utilities/Algorithm.hpp"
+#include "Utilities/TMPL.hpp"
+
+/// \cond
+template <size_t>
+class ElementId;
+/// \endcond
+
+namespace amr::Criteria {
+
+namespace Constraints_detail {
+
+/*!
+ * \brief Computes a measure of the constraints in each logical direction of the
+ * grid
+ *
+ * \requires `constraints_tensor` to be a tensor of rank 2 or higher. The first
+ * index must be a lower spatial index that originates from a derivative.
+ *
+ * We follow \cite Szilagyi2014fna, Eq. (62)-(64) to compute:
+ *
+ * \begin{align}
+ * \Epsilon_\hat{k} = \frac{1}{N_\hat{k}} \sqrt{\sum_{a\ldots} \left(\sum_{i}
+ *   \frac{\partial x^i}{\partial x^\hat{k}} C_{ia\ldots}\right)^2} \\
+ * N_\hat{k} = \sqrt{\sum_{i} \left(\frac{\partial x^i}{\partial x^\hat{k}}
+ *   \right)^2}
+ * \end{align}
+ *
+ * This transform the first lower spatial index of the tensor to the
+ * element-logical frame, then takes an L2 norm over all remaining indices.
+ */
+template <size_t Dim, typename TensorType>
+void logical_constraints(
+    gsl::not_null<tnsr::i<DataVector, Dim, Frame::ElementLogical>*> result,
+    gsl::not_null<DataVector*> buffer, const TensorType& constraints_tensor,
+    const Jacobian<DataVector, Dim, Frame::ElementLogical, Frame::Inertial>&
+        jacobian);
+
+/*!
+ * \brief Apply the AMR criterion to one of the constraints
+ *
+ * The `result` is the current decision in each dimension based on the previous
+ * constraints. This function will update the flags if necessary. It takes
+ * the "max" of the current and new flags, where the "highest" flag is
+ * `Flag::IncreaseResolution`, followed by `Flag::DoNothing`, and then
+ * `Flag::DecreaseResolution`.
+ */
+template <size_t Dim>
+void max_over_components(
+    gsl::not_null<std::array<Flag, Dim>*> result,
+    const tnsr::i<DataVector, Dim, Frame::ElementLogical>& logical_constraints,
+    double abs_target, double coarsening_factor);
+
+}  // namespace Constraints_detail
+
+/*!
+ * \brief Refine the grid towards the target constraint violation
+ *
+ * - If any constraint is above the target value, the element will be p-refined.
+ * - If all constraints are below the target times the "coarsening factor" the
+ *   element will be p-coarsened.
+ *
+ * This criterion is based on Sec. 6.1.4 in \cite Szilagyi2014fna .
+ *
+ * If the coarsening factor turns out to be hard to choose, then we can try to
+ * eliminate it by projecting the variables to a lower polynomial order before
+ * computing constraints, or something like that.
+ *
+ * \tparam Dim Spatial dimension of the grid
+ * \tparam TensorTags List of tags of the constraints to be monitored. These
+ * must be tensors of rank 2 or higher. The first index must be a lower spatial
+ * index that originates from a derivative.
+ */
+template <size_t Dim, typename TensorTags>
```

Comment:
Are these going to be evaluated using the `ObservationBox`? I don't want to store the 2 & 4-index constraints in the DataBox. We don't have enough RAM for that.